pkg/plugins/middlewares.go

```diff
@@ -34,30 +39,73 @@ func (b Builder) Build(pName string, config map[string]interface{}, middlewareNa
 type middlewareBuilder struct {
 	fnNew          reflect.Value
 	fnCreateConfig reflect.Value
+	pluginType     string
+	wasmPath       string
 }
 
-func newMiddlewareBuilder(i *interp.Interpreter, basePkg, imp string) (*middlewareBuilder, error) {
+func findPluginType(wasmPath string) string {
+	if wasmPath != "" {
+		return PluginTypeWASM
+	}
+	return PluginTypeYaegi
+}
+
+func newMiddlewareBuilder(i *interp.Interpreter, basePkg, imp, wasmPath string) (*middlewareBuilder, error) {
+	builder := &middlewareBuilder{
+		wasmPath:   wasmPath,
+		pluginType: findPluginType(wasmPath),
+	}
+
+	if builder.pluginType == PluginTypeWASM {
+		return builder, nil
+	}
+
 	if basePkg == "" {
 		basePkg = strings.ReplaceAll(path.Base(imp), "-", "_")
 	}
+	var err error
 
-	fnNew, err := i.Eval(basePkg + `.New`)
+	builder.fnNew, err = i.Eval(basePkg + `.New`)
 	if err != nil {
 		return nil, fmt.Errorf("failed to eval New: %w", err)
 	}
 
-	fnCreateConfig, err := i.Eval(basePkg + `.CreateConfig`)
+	builder.fnCreateConfig, err = i.Eval(basePkg + `.CreateConfig`)
 	if err != nil {
 		return nil, fmt.Errorf("failed to eval CreateConfig: %w", err)
 	}
-
-	return &middlewareBuilder{
-		fnNew:          fnNew,
-		fnCreateConfig: fnCreateConfig,
-	}, nil
+	return builder, nil
 }
 
 func (p middlewareBuilder) newHandler(ctx context.Context, next http.Handler, cfg reflect.Value, middlewareName string) (http.Handler, error) {
+	if p.pluginType == PluginTypeWASM {
+		code, err := os.ReadFile(p.wasmPath)
+		if err != nil {
+			return nil, err
+		}
+		logger := middlewares.GetLogger(ctx, middlewareName, PluginTypeWASM)
+		opts := []handler.Option{
+			handler.Logger(initWasmLogger(logger)),
+		}
+		i := cfg.Interface()
+		if i != nil {
+			config, ok := i.(map[string]interface{})
+			if !ok {
+				return nil, fmt.Errorf("could not type assert config: %T", i)
```

Comment:
I wonder if this type assertion is required. I think the marshaling will do also some sort of type assertion,

Reply:
not sure, but it does not cost much to double check